src/datasets/metric.py

```diff
@@ -336,7 +336,8 @@ def _finalize(self):
         from all the nodes if main node or all_process is True.
         """
         if self.writer is not None:
-            self.writer.finalize()
+            with self.writer as writer:
+                writer.finalize()
```

Comment:
The usage of the `with` statement here can make the user think that the writer opens and creates a file here.
This is not the case since finalize just writes down the latest examples and close the file.

It would be prettier to use .close() instead of a with statement here, to remove this ambiguity.

The with statement should be used to define the start and end point of the usage of the writer (file opened until closed).
However in this case the writer is an attribute of the Metric object and so there's no code block with a defined start/end.
In this case I think it's better to use .close() instead. You can even do try/except to make sure that close() is called even if finalize fails.

It would makes things clearer IMO.

Reply:
@lhoestq, I think the main point is in your sentence: 
> You can even do try/except to make sure that close() is called even if finalize fails.
Exactly! This is why I decided to open this PR: an exception was raised, and the resource was not closed. If an exception is raised within .finalize(), the file will no be closed. So that the solution would be: try... except... finally: close().

And that is exactly the use case for `with`: use a context manager instead of try... except... finally: close().
See: https://docs.python.org/3/reference/compound_stmts.html#the-with-statement
> This allows common try…except…finally usage patterns to be encapsulated for convenient reuse.

Reply:
In this case maybe you can you add in the writer's docstring that the `with` can also be used only when calling `finalize` ?
As a user I'd expect the `with` statement of a `Writer` class to be necessarily used from start of writing till closing, but here you can write and then use the `with` statement to close. Let me know if that makes sense. 

That's definitely not a big deal but I'd rather have things said explicitly :)

Reply:
Sounds good !